File: 2024/2/2.py

import sys

def sign(n):
    if n==0:
        return 0
    return n/abs(n)

def isSafe(report):
    bs = sign(report[1] - report[0])
    for i in range(1, len(report)):
        last, cur = report[i-1], report[i]
        dif = cur-last
        if sign(dif) != bs or dif == 0 or abs(dif) > 3:
            return False
    return True

safe = 0
for line in sys.stdin.readlines():
    report = [int(n) for n in line.strip().split(" ")]
    if isSafe(report):
        safe += 1
    else:
        for i in range(0, len(report)):
            copy = report.copy()
            copy.pop(i)
            if isSafe(copy):
                safe += 1
                break
print(safe)

File: 2024/2/2.lua
local safe = 0

local function nsplit(line)
  local w = {}
  for t in line:gmatch("[^ ]+") do
    w[#w+1] = tonumber(t)
  end
  return w
end

local function sign(n)
  if n == 0 then return 1 end
  return n/math.abs(n)
end

local function cp(t)
  local n={}
  for i=1, #t do n[i] = t[i] end
  return n
end

local function testSafety(t, r)
  if r then table.remove(t, r) end
  local bs = sign(t[2] - t[1])
  for i=2, #t do
    local cur, last = t[i], t[i-1]
    local dif = cur - last
    if sign(dif) ~= bs or math.abs(dif) == 0 or math.abs(dif) > 3 then
      return false
    end
  end
  return true
end

for line in io.lines() do
  local num = nsplit(line)
  if testSafety(num) then
    safe = safe + 1
  else
    for i=1, #num do
      if testSafety(cp(num), i) then
        safe = safe + 1
        break
      end
    end
  end
end

print(safe)
